File: activeaero/apogee_prediction.py
from time import sleep
import time
import math
from math import cos
from math import sin
from ulab import numpy as np


# Constants to be set prior to flight
mass = 2.132     # Mass of rocket + empty motor, needs to be calculated prior to each flight
area = 0.0049483 # Cross-sectional area of tube, in m^2
rho = 1.22      # Air density, going to be constant at low altitudes but will need to be calculated for IREC
k = 0.5 * rho * 0.58 * area # Wind resistance

# Following lines need to be replaced with whatever the code uses to gather sensor data


t = 0       # Needs to be set so pressing a push button will set this off

# Following code is running in "armed" state

class ApogeePrediction():
    def __init__(self):
        self.looptime = round(time.monotonic_ns() / 10**9 * 10**3)
        self.lastalt = 0
        self.curaltm = 0
        self.curvelacc = 0
        self.lastvelpres = 0

    def predict_apogee(self, curalt, az):#, ax, ay, opit, orol):
        lasttime = self.looptime
        self.looptime = round(time.monotonic_ns() / 10**9 * 10**3)
        timedif = (self.looptime - lasttime)/1000
        curvelpres = ((curalt - self.lastalt)/timedif + self.lastvelpres)/2
        self.lastalt = curalt

        # azvect = az*sin(opit) + ay*cos(opit)*sin(orol) + ax*cos(opit)*cos(orol)     # True vertical acceleration, if the code is set up correctly, any orientation should register 1g of acceleration
        azvect = az
        apaz = (azvect-1)                                                      # True vertical acceleration, removes normal force from accelerometer
        if apaz < 0.01 or apaz > -0.01:
            apaz = 0
        lastvelacc = self.curvelacc
        self.curvelacc = apaz * timedif + lastvelacc                                           # Calculates the vertical velocity

        appredone = (mass / (2*k) * np.log((mass*9.8+self.curvelacc**2)/(mass*9.8))) + curalt   # First apogee prediction, uses velocity from accelerometer
        appredtwo = (mass / (2*k) * np.log((mass*9.8+curvelpres**2)/(mass*9.8))) + curalt  # Second apogee prediction, uses velocity from altitude
        
        self.lastvelpres = curvelpres
        
        return appredone, appredtwo


    # STILL NEEDED:
    # Sensors need to be all added in
    # Data needs to be recorded to some file, all sensor data, curalt, self.curvelacc, curvelpres, appredone, appredtwo
    # With data recording, I'd like to use the time interval stuff again to prevent clogging of files
    # Push button code to 'arm' the system, I'm also thinking we put a little light in to check if it's armed too

File: activeaero/apogeeprediction.cpp
#include <Arduino.h>

#include "apogeeprediction.h"
#include "math.h"

ApogeePrediction::ApogeePrediction(double rocketMass, double dragCoefficent, double crossArea, double targetApogee) : rocketMass(rocketMass), dragCoefficent(dragCoefficent), crossArea(crossArea), targetApogee(targetApogee) {
    currentVelocity = 0;
    lastRecTime = 0;
    predApogee = 0;
}

void ApogeePrediction::eulerFromQuaternion(double *euler, double x, double y, double z, double w) {
    double t0 = +2.0 * (w * x + y * z);
    double t1 = +1.0 - 2.0 * (x * x + y * y);
    double roll_x = atan2(t0, t1);
    double t2 = +2.0 * (w * y - z * x);
    t2 = + t2 > +1.0 ? 1.0 : t2;
    t2 = t2 < -1.0 ? -1.0 : t2;
    double pitch_y = asin(t2);
    double t3 = +2.0 * (w * z + x * y);
    double t4 = +1.0 - 2.0 * (y * y + z * z);
    double yaw_z = atan2(t3, t4);

    euler[0] = roll_x;
    euler[1] = pitch_y;
    euler[2] = yaw_z;

    // Convert to degrees if needed
    // roll_x = roll_x * 180 / 3.14159
    // pitch_y = pitch_y * 180 / 3.14159
    // yaw_z = yaw_z * 180 / 3.14159
}

void ApogeePrediction::calcVelocity(double acceleration) {
    double curTime = micros();
    double deltaTime = curTime - lastRecTime;
    currentVelocity += acceleration * deltaTime;
    lastRecTime = curTime;
}

double ApogeePrediction::predictApogee(double* acceleration, double* orientation, double pressure, double temperature, double altitude) {
    double euler[3] = {0, 0, 0};
    eulerFromQuaternion(euler, orientation[0], orientation[1], orientation[2], orientation[3]);  // Convert to radians
    double azvect = -acceleration[0] * sin(euler[1]) + acceleration[1]*cos(euler[1])*sin(euler[0]) + acceleration[2] * cos(euler[1])*cos(euler[0]); // vertical acceleration component, from the bno08x
    calcVelocity(azvect);
    
    double rho = pressure*100/(287.058*(temperature+273.15)); // Get dry air density
    double k = 0.5*rho*dragCoefficent*crossArea;
    double predApogee = ((rocketMass/(2*k))*log((rocketMass*9.81 + k*pow(currentVelocity,2))/(rocketMass*9.81))+altitude); // Apogee prediction in meters

    return predApogee;
}